mods/parsoid.js

```diff
@@ -144,64 +143,140 @@ PSP.transformRevision = function (restbase, req, from, to) {
+                    'content-type': res.headers['content-type']
+                },
+                body: res.body
+            };
         });
     }
 
-    return Promise.all([ get('html'), get('wikitext'), get('data-parsoid') ])
-    .then(function () {
+    // Get the revision info just to make sure we have access
+    return self.getRevisionInfo(restbase, req)
+    .then(function(revInfo) {
+        return Promise.props({
+            html: get('html'),
+            // wikitext: get('wikitext'),
+            'data-parsoid': get('data-parsoid')
+        });
+    })
+    .then(function (original) {
+        original.revid = rp.revision;
         var body2 = {
-            original: fromStorage
+            original: original
         };
-        body2[from] = req.body;
-        return restbase.post({
-            uri: new URI([rp.domain,'sys','parsoid','transform',from,'to',to]),
+        body2[from] = req.body[from];
+        var path = [rp.domain,'sys','parsoid','transform',from,'to',to];
+        if (rp.title) {
+            path.push(rp.title);
+            if (rp.revision) {
+                path.push(rp.revision);
+            }
+        }
+        var newReq = {
+            uri: new URI(path),
+            params: req.params,
             headers: { 'content-type': 'application/json' },
             body: body2
-        });
+        };
+        return self.callParsoidTransform(restbase, newReq, from, to);
     });
 
 };
 
+PSP.callParsoidTransform = function callParsoidTransform (restbase, req, from, to) {
+    var rp = req.params;
+    // Parsoid currently spells 'wikitext' as 'wt'
+    var parsoidTo = to;
+    if (to === 'wikitext') {
+        parsoidTo = 'wt';
+    } else if (to === 'html') {
+        // Retrieve pagebundle whenever we want HTML
+        parsoidTo = 'pagebundle';
+    }
+
+
+    var parsoidExtras = [];
+    if (rp.title) {
+        parsoidExtras.push(rp.title);
+    } else {
+        // fake title to avoid Parsoid error: <400/No title or wikitext was provided>
+        parsoidExtras.push('Main_Page');
+    }
+    if (rp.revision) {
+        parsoidExtras.push(rp.revision);
+    }
+    var parsoidExtraPath = parsoidExtras.map(encodeURIComponent).join('/');
+    if (parsoidExtraPath) { parsoidExtraPath = '/' + parsoidExtraPath; }
+
+    var domain = rp.domain;
+    // Re-map test domain
+    if (domain === 'en.wikipedia.test.local') { domain = 'en.wikipedia.org'; }
+    var parsoidReq = {
+        uri: this.parsoidHost + '/v2/' + domain + '/'
+            + parsoidTo + parsoidExtraPath,
+        headers: { 'content-type': 'application/json' },
+        body: req.body
+    };
+    console.log(JSON.stringify(parsoidReq, null, 2));
+    return restbase.post(parsoidReq);
+};
+
+/**
+ * Cheap body.innerHTML extraction.
+ *
+ * This is safe as we know that the HTML we are receiving from Parsoid is
+ * serialized as XML.
+ */
+function cheapBodyInnerHTML(html) {
+    var match = /<body[^>]*>(.*)<\/body>/.exec(html);
```

Comment:
I believe a body can span multiple lines. If so, the correct way would probably be:

```javascript
var re = new RegExp('/<body[^>]*>([\\s\\S]*)<\\/body>/', 'gm');
var match = re.exec(html);
```